pkg/kv-cache/kv-block/in-memory.go

```diff
@@ -0,0 +1,200 @@
+	"k8s.io/apimachinery/pkg/util/sets"
+	"k8s.io/klog/v2"
+)
+
+const (
+	defaultInMemoryIndexSize = 1e8 // TODO: change to memory-size based configuration
+	defaultPodCacheSize      = 10  // number of pods per key
+)
+
+// InMemoryIndexConfig holds the configuration for the InMemoryIndex.
+type InMemoryIndexConfig struct {
+	// Size is the maximum number of keys that can be stored in the index.
+	Size int
+	// PodCacheSize is the maximum number of pod entries per key.
+	PodCacheSize int
+}
+
+// DefaultInMemoryIndexConfig returns a default configuration for the InMemoryIndex.
+func DefaultInMemoryIndexConfig() *InMemoryIndexConfig {
+	return &InMemoryIndexConfig{
+		Size:         defaultInMemoryIndexSize,
+		PodCacheSize: defaultPodCacheSize,
+	}
+}
+
+// NewInMemoryIndex creates a new InMemoryIndex instance.
+func NewInMemoryIndex(cfg *InMemoryIndexConfig) (*InMemoryIndex, error) {
+	if cfg == nil {
+		cfg = DefaultInMemoryIndexConfig()
+	}
+
+	cache, err := lru.New[Key, *PodCache](cfg.Size)
+	if err != nil {
+		return nil, fmt.Errorf("failed to initialize in-memory index: %w", err)
+	}
+
+	return &InMemoryIndex{
+		data:         cache,
+		podCacheSize: cfg.PodCacheSize,
+	}, nil
+}
+
+// InMemoryIndex is an in-memory implementation of the Index interface.
+type InMemoryIndex struct {
+	// data holds the mapping of keys to sets of pod identifiers.
+	data *lru.Cache[Key, *PodCache]
+	// podCacheSize is the maximum number of pod entries per key.
+	podCacheSize int
+}
+
+var _ Index = &InMemoryIndex{}
+
+// PodCache represents a cache for pod entries.
+type PodCache struct {
+	// cache is an LRU cache that maps PodEntry to their last access time.
+	// thread-safe.
+	cache *lru.Cache[PodEntry, time.Time]
+}
+
+// Lookup receives a list of keys and a set of pod identifiers,
+// and retrieves the filtered pods associated with those keys.
+// The filtering is done based on the pod identifiers provided.
+// If the podIdentifierSet is empty, all pods are returned.
+//
+// It returns:
+// 1. A slice of the hit keys.
+// 2. A map where the keys are those in (1) and the values are pod-identifiers.
+// 3. An error if any occurred during the operation.
+func (m *InMemoryIndex) Lookup(ctx context.Context, keys []Key,
+	podIdentifierSet sets.Set[string],
+) ([]Key, map[Key][]string, error) {
+	if len(keys) == 0 {
+		return nil, nil, fmt.Errorf("no keys provided for lookup")
+	}
+
+	traceLogger := klog.FromContext(ctx).V(5).WithName("kvblock.InMemoryIndex.Lookup")
+
+	podsPerKey := make(map[Key][]string)
+	highestHitIdx := 0
+
+	for idx, key := range keys {
+		if pods, found := m.data.Get(key); found { //nolint:nestif // TODO: can this be optimized?
+			if pods == nil || pods.cache.Len() == 0 {
+				traceLogger.Info("no pods found for key, cutting search", "key", key.String())
+				return keys[:idx], podsPerKey, nil // early stop since prefix-chain breaks here
+			}
+
+			highestHitIdx = idx
+
+			if podIdentifierSet.Len() == 0 {
```

Comment:
Unless I misunderstand the code or intent, can this test be moved outside the loop?
Does `podIdentifierSet.Len()` ever change inside the loop?

Reply:
This can be more readable (trivially more performant?) if done similarly to the redis index:
`filterPods := len(podIdentifierSet) > 0 // predicate for filtering`. But the check cannot be avoided - or at least I'm missing the obvious way.

Reply:
the check is fine. I just did not think it needs to be inside the loop as its evaluation is independent of loop variables and conditions.